### ISSUE TITLE
CLAP: blocking config bug (layer_norm_eps 1e-5 vs 1e-12) + audit findings

### ISSUE THREAD

**Issue by SRjoeee** [findit-studio/mlxrs#365]
# CLAP: deep audit findings — blocking config bug + testing gaps

## Summary

A 2-round adversarial audit (architecture, config, mel frontend, audio tower HTSAT, text tower RoBERTa, projection, E2E parity) of the CLAP-HTSAT-unfused port against **HF `transformers` `modeling_clap.py`** and **Findit-AI/textclap** (Rust ONNX reference).

**Result: 1 blocking bug, 1 dormant bug, zero correctness bugs in the inference path.**

---

## 1. BLOCKING BUG — `text_config.layer_norm_eps` wrong default and pin

**Location:** `config.rs:320-322` (default), `config.rs:546` (pin)

**Evidence chain:**

1. **Upstream `laion/clap-htsat-unfused/config.json`:**
   ```json
   "text_config": { "layer_norm_eps": 1e-12 }
   ```
   (RoBERTa default — `1e-12`, not `1e-5`)

2. **mlxrs default** (`config.rs:320-322`):
   ```rust
   fn default_text_layer_norm_eps() -> f64 {
     1e-5  // WRONG: 7 orders of magnitude too large
   }
   ```

3. **mlxrs validate() pin** (`config.rs:546`):
   ```rust
   pin_f64("ClapTextConfig: layer_norm_eps", self.layer_norm_eps, 1e-5)?;
   //                                              pin: 1e-5 — WRONG
   ```

4. **`pin_f64` uses exact equality** (`model_validation/mod.rs:189`):
   ```rust
   if actual != expected { return Err(...) }
   ```

**Impact:**
- **Loading fails**: The real checkpoint's `config.json` has `1e-12`; `validate()` rejects it because `1e-12 != 1e-5`.
- **Wrong default**: A config.json omitting the field silently uses `1e-5` instead of `1e-12`.
- **Numerical impact**: Every RoBERTa LayerNorm computes `sqrt(var + eps)`. With eps=1e-5 instead of 1e-12, the normalization floor is 10,000,000x larger. When variance is small, this corrupts text embeddings and degrades contrastive alignment.

**Fix (3 changes):**
1. `config.rs:320`: Change default from `1e-5` to `1e-12`
2. `config.rs:546`: Change pin from `1e-5` to `1e-12`
3. `config.rs:173`: Update doc comment from `1e-5` to `1e-12`

**Note:** `audio_config.layer_norm_eps = 1e-5` is correct — only the text side is wrong.

---

## 2. DORMANT BUG — E2E test attention mask dtype mismatch

**Location:** `tests/clap_e2e_parity.rs:298,313`

The `tokenize_labels_batch` helper creates attention masks as `Vec<i32>` / `Array::from_slice::<i32>`, but `build_additive_mask` (`text.rs:555-556`) compares against `Array::full::<f32>(&(1,), 0.0)`. Whether MLX's `not_equal` handles mixed i32/f32 depends on type promotion.

The unit tests (`text/tests.rs:469-474`) correctly use `Array::from_slice::<f32>` for masks.

**Impact:** Currently dormant (`#[ignore]` pending model conversion). Would surface as a test failure or wrong results when the E2E tests are enabled.

**Fix:** Change `Vec<i32>` to `Vec<f32>` in `tokenize_labels_batch`.

---

## 3. Architecture verification — all MATCH

| Component | Status | Details |
|-----------|--------|---------|
| HTSAT 4-stage Swin (depths [2,2,6,2]) | MATCH | Stage dim doubling 96→768, resolution halving |
| Window attention + relative-position bias | MATCH | Table gather + pre-softmax additive bias |
| SW-MSA mask (region labels) | MATCH | Closed-form band(0,1,2) × 3 + wband |
| Cyclic shift (roll_axes) | MATCH | shift=0 skip, shift=win/2 correct roll |
| PatchMerging (2×2, no-bias reduction) | MATCH | Concat order pinned by oracle |
| PatchEmbed Conv2d (4×4, stride 4) | MATCH | NHWC weight transpose correct |
| reshape_mel2img (bicubic + freq_ratio fold) | MATCH | Two-step interpolate + reshape + permute |
| Token-semantic mean pool | MATCH | Mean-invariant to spatial rearrangement |
| RoBERTa text encoder | MATCH | Post-norm, exact GELU, CLS pooling |
| Position ID offset (cumsum) | MATCH | Oracle: [0,5,9,1,1] → [2,3,4,1,1] |
| Projection (Linear→ReLU→Linear) | MATCH | NOT GELU — pinned `projection_hidden_act="relu"` |
| L2 normalization | MATCH | eps=1e-9 (immaterial for real embeddings) |
| Zero-shot classify (cosine + sort) | MATCH | L2-normed dot product, descending stable sort |
| Mel frontend (N_FFT=1024, HOP=480, N_MELS=64) | MATCH | All DSP constants match textclap |

---

## 4. Benign divergences (documented, not bugs)

| Divergence | mlxrs | Reference | Impact |
|------------|-------|-----------|--------|
| STFT precision | f32 (MLX rfft) | f64 (rustfft/numpy) | ~4.4e-4 dB drift, well-tested |
| Filterbank layout | (n_freqs, n_mels) transposed | (n_mels, n_freqs) | Mathematically equivalent |
| Mel output layout | (1,1,T,M) 4D | (T,M) 2D | HF layout match, intentional |
| Window storage | f32 | f64 | Sub-1e-6 error |
| DropPath | omitted | present | Inference no-op (identity at eval) |

---

## 5. Feature gaps vs textclap

| Feature | textclap | mlxrs | Priority |
|---------|----------|-------|----------|
| Chunking API (`embed_chunked` for >10s audio) | ✅ `ChunkingOptions` | ❌ Missing | MEDIUM |
| Fused path (`laion/clap-htsat-fused`) | ✅ (ONNX black-box) | ❌ Rejected by validate | LOW |

---

## 6. mlxrs advantages over textclap

| Advantage | Details |
|-----------|---------|
| Quantization support | `PerLayerQuantization` for both towers + projections |
| validate() pinning | Every architecture field pinned to checkpoint value |
| Runtime shape checks | `take_shaped`, `checked_mul`, `require_divisible` |
| Oracle tests | Mel filterbank, relative-position index, PatchMerging concat order |

---

## 7. E2E parity test status

The `clap_e2e_parity.rs` tests are well-designed (cosine tolerance, rank agreement, checkpoint-absent guard) but **currently dormant**:
- All 3 tests are `#[ignore]`
- Blocked on `pytorch_model.bin → model.safetensors` conversion
- Golden data from textclap's int8-quantized ONNX
- Tolerance: `1 - cosine <= 1e-2` (embedding), `SCORE_BAND = 5e-2` (classify)

---

## Recommendation

- **Fix #1 immediately** — the `layer_norm_eps` bug blocks loading any real CLAP checkpoint.
- **Fix #2** when enabling E2E tests — the dtype mismatch will cause failures.
- Track chunking API as a feature request.

**Comment by SRjoeee** [comment]
## Round 3 findings (追加)

### NEW BUG 3: Text L2 normalization eps mismatch

**Location:** 

The audio path uses  with  (matching PyTorch  default).

The text path uses bare  which falls through to .

Both paths feed into the same contrastive cosine space. HF CLAP uses  for BOTH towers (default eps=1e-12).

**Impact:** Negligible for f32 non-degenerate vectors. Would matter for f16 activations or near-zero edge cases.

**Fix:**  — change `l2_normalize(&projected)` to `l2_normalize_eps(&projected, SWIFT_L2_EPS)`.

---

### DESIGN CONCERN 4: HF ClapTextPooler omitted

mlxrs does plain CLS pooling (just take first token). HF's  uses  which goes through a  (dense + tanh).

This follows the **original LAION CLAP code** (which does  without pooler), not HF's port. The checkpoint's  keys are silently dropped by sanitize.

**Risk:** If the pooler has non-trivial trained weights, mlxrs text embeddings will silently diverge from HF's output. The E2E parity test against textclap (which also skips the pooler via ONNX export) would not catch this.

**Action:** Verify against the real checkpoint whether  exists and is non-trivial. If so, either port the pooler or document the divergence.

---

### Performance note: shifted_window_mask recomputation

: The SW-MSA mask is recomputed every forward for shifted blocks, but depends only on fixed per-stage dimensions. Could be cached at construction time like RelativePositionBias already is. Low priority — the mask is small (~64K floats for window=8).

---

### All other paths verified correct

- Quantization path ✅
- BatchNorm precombination ✅
- Bicubic interpolation (align_corners=True) ✅
- Mel pipeline constants ✅
- logit_scale not applied at inference ✅
- ~77 transposes/audio, ~74 transposes/text (standard MLX pattern) ✅
- Mel computed on GPU ✅
- No unnecessary dtype promotions ✅

**Comment by SRjoeee** [comment]
## Round 3 findings (追加)

### NEW BUG 3: Text L2 normalization eps mismatch

**Location:** `text.rs:452`

The audio path uses `l2_normalize_eps(&projected, SWIFT_L2_EPS)` with `SWIFT_L2_EPS = 1e-12` (matching PyTorch `F.normalize` default).

The text path uses bare `l2_normalize(&projected)` which falls through to `DEFAULT_NORMALIZE_EPS = 1e-9`.

Both paths feed into the same contrastive cosine space. HF CLAP uses `F.normalize(x, dim=-1)` for BOTH towers (default eps=1e-12).

**Impact:** Negligible for f32 non-degenerate vectors. Would matter for f16 activations or near-zero edge cases.

**Fix:** `text.rs:452` — change `l2_normalize(&projected)` to `l2_normalize_eps(&projected, SWIFT_L2_EPS)`.

---

### DESIGN CONCERN 4: HF ClapTextPooler omitted

mlxrs does plain CLS pooling (just take first token). HF's `ClapModel.get_text_features` uses `text_outputs.pooler_output` which goes through a `ClapTextPooler` (dense + tanh).

This follows the **original LAION CLAP code** (which does `x = x[:, 0, :]` without pooler), not HF's port. The checkpoint's `text_model.pooler.dense.{weight,bias}` keys are silently dropped by sanitize.

**Risk:** If the pooler has non-trivial trained weights, mlxrs text embeddings will silently diverge from HF's output. The E2E parity test against textclap (which also skips the pooler via ONNX export) would not catch this.

**Action:** Verify against the real checkpoint whether `text_model.pooler.dense.weight` exists and is non-trivial. If so, either port the pooler or document the divergence.

---

### Performance note: shifted_window_mask recomputation

`shared.rs:1598-1607`: The SW-MSA mask is recomputed every forward for shifted blocks, but depends only on fixed per-stage dimensions. Could be cached at construction time like RelativePositionBias already is. Low priority — the mask is small (~64K floats for window=8).

---

### All other paths verified correct

- Quantization path ✅
- BatchNorm precombination ✅
- Bicubic interpolation (align_corners=True) ✅
- Mel pipeline constants ✅
- logit_scale not applied at inference ✅
- ~77 transposes/audio, ~74 transposes/text (standard MLX pattern) ✅
- Mel computed on GPU ✅
- No unnecessary dtype promotions ✅

**Comment by SRjoeee** [comment]
## Round 4 findings

### Verification ✅

All 3 fixes from #368 verified correct:

| Fix | Status | Evidence |
|-----|--------|----------|
| layer_norm_eps 1e-12 | ✅ | default, pin, doc comment, tests all updated |
| ClapTextPooler (dense+tanh) | ✅ | Correct struct, loaded from pooler.dense.*, applied after CLS before projection |
| L2 normalize eps SWIFT_L2_EPS | ✅ | Both text and audio paths use 1e-12, matching HF F.normalize default |

---

### Performance Optimization Opportunities

**P1. HIGH — SW-MSA mask recomputed every forward pass**

`shared.rs:1598-1612`: 5 out of 12 Swin blocks are shifted (odd-indexed in stages 0-2). Each shifted block calls `shifted_window_mask()` which builds a CPU Vec<f32>, converts to Array, and reshapes — every single forward. The code comment explicitly acknowledges this could be cached.

**Suggestion:** Thread `(height, width)` into `SwinBlock::from_weights` (already available per-stage). Build the mask once at construction, identical to how `RelativePositionBias` is already precomputed. Eliminates 5 CPU→GPU transfers per forward. Memory cost: ~1 MB (negligible).

**P2. MEDIUM — Q/K/V projected as 3 separate matmuls**

`shared.rs:1318-1320`: Three separate `QuantLinear::forward` calls = 3 kernel launches + 3 intermediate allocations.

**Suggestion:** Fuse into single `(dim, 3·dim)` weight matrix at load time:
```
W_qkv = concat([W_q, W_k, W_v], axis=1)  // load-time only
qkv = matmul(x, W_qkv)                     // single kernel
q, k, v = split(qkv, [dim, dim, dim])      // view, no copy
```
Reduces kernel launches from 3→1. Standard Swin optimization (used in timm). Same applies to all 12 SwinBlocks.

**P3. MEDIUM — RelativePositionBias::additive() recomputes expand+astype every forward**

`shared.rs:1107-1110`: Called 12 times per forward (once per SwinBlock). `expand_dims` produces a view (cheap) but `astype` may trigger a copy if dtype differs.

**Suggestion:** Cache the expanded, dtype-converted version at construction. Store as `(1, heads, win², win²)` in the activation dtype.

**P4. MEDIUM — EncoderBatchNorm casts scale/shift every forward**

`audio.rs:179-184`: `astype(&self.scale, dtype)?` on every forward. If both are f32, this is a no-op copy.

**Suggestion:** Cache the activation-dtype versions at construction. Low priority since BN is called once per forward (not per-block).

---

### New Issues Found

**ISSUE 5: `logit_scale_a` / `logit_scale_t` silently orphaned in weight map**

`mod.rs:447-448`: The sanitize function keeps `logit_scale_a` and `logit_scale_t` verbatim in the output map. After `from_weights` splits the map per tower, these keys remain unconsumed. No warning is logged.

**Suggestion:** Either drop them explicitly in sanitize (like `position_ids`), or log a warning for unconsumed keys. Inconsistent with the strict shape-pinning discipline elsewhere.

**ISSUE 6: E2E parity tests have never run**

`tests/clap_e2e_parity.rs`: All 3 tests are `#[ignore]` AND require `model.safetensors` which doesn't exist (HF ships `pytorch_model.bin` only). The golden parity has never been validated against the real checkpoint.

**Suggestion:** Convert `pytorch_model.bin → model.safetensors` once and check it in, or document this as a known gap.

**ISSUE 7: Minor test coverage gaps**

- Batch `embed_audio` with B > 1 (all tests use B=1)
- `embed_text` with all-pad rows (what does the pooler produce?)
- Empty waveform handling
- `classify` with k > num_labels

---

### All correctness paths verified ✅

- reshape_mel2img fold arithmetic ✅
- PatchMerging concat order ✅
- Relative-position bias table gather ✅
- SW-MSA mask generation ✅
- Token-semantic mean pool ✅
- Bicubic interpolation two-pass ✅
- BatchNorm precombination ✅
- Quantization path ✅
- Cross-model normalization consistency ✅

**Comment by SRjoeee** [comment]
## Final consolidated findings (4-round internal audit + independent Claude Code review)

### Fix verification ✅ (all 3 fixes confirmed correct by both reviewers)

| Fix | Status | Evidence |
|-----|--------|----------|
| layer_norm_eps 1e-12 | ✅ | Default, pin, doc, tests all correct. Verified against real config.json. |
| ClapTextPooler (dense+tanh) | ✅ | Correct struct, loaded from pooler.dense.*, applied after CLS before projection. Perturbation test proves it's in the forward path. |
| L2 normalize eps SWIFT_L2_EPS | ✅ | Both text and audio paths use 1e-12, matching HF F.normalize default. |

All activations verified correct: exact erf GELU in towers, tanh in pooler, ReLU in projections.

---

### Actionable findings (cross-validated by both reviewers)

| # | Severity | Finding | Status |
|---|----------|---------|--------|
| 1 | **MEDIUM** | SW-MSA shifted_window_mask rebuilt every forward (5 blocks × 1 call). Depends only on fixed per-stage dims. Could be cached at construction like RelativePositionBias. | Tracked #365, acknowledged in code comment |
| 2 | **HIGH** | No end-to-end numerical parity test against real checkpoint. Tower/projection/pooler assembly is verified only by shape tests with random weights. The individual primitives are oracle-pinned, but composition is unverified. | Known gap, deferred to separate PR |

### LOW findings (polish, from both reviewers)

| # | Finding | Source |
|---|---------|--------|
| 3 | `logit_scale_a/t` silently orphaned in weight map (not consumed, no warning) | Both |
| 4 | Pad mask uses `-inf` vs HF `finfo(dtype).min` (numerically negligible) | Claude Code |
| 5 | `patch_stride` / `audio hidden_act` not modeled (safe via patch_size/GELU pins) | Claude Code |
| 6 | `reshape_mel2img` 2nd interpolation uses `cur_time` vs HF `stale time_length` (dead code for pinned config) | Claude Code |
| 7 | `from_weights_quantized` no leftover-keys check | Both |
| 8 | B>1 shifted-window path untested (correct by analysis) | Both |
| 9 | f32 STFT vs f64 reference (~4.4e-4 dB drift, documented) | Both |
| 10 | classify returns cosine not logit_scale·cosine (ranking identical, documented) | Both |

### Performance claims — adversarial verification results

| Claim | Verdict | Reason |
|-------|---------|--------|
| SW-MSA mask recomputed (P1) | **CONFIRMED** | Real recomputation, ~6 calls/forward, small arrays |
| Q/K/V fusion (P2) | **QUALIFY** | Same shapes but QuantLinear abstraction makes fusion complex |
| RelativePositionBias expand+astype (P3) | **REFUTE** | expand_dims is lazy view, astype zero-cost when dtype matches |
| BatchNorm astype (P4) | **REFUTE** | astype zero-cost when dtype matches, BN called once per forward |

### Bottom line

The implementation is correct. All 3 reported bugs are properly fixed. The only actionable optimization is caching the SW-MSA mask (P1). The highest-priority gap is the end-to-end numerical parity test (HIGH).
